sympy/solvers/solveset.py

```diff
@@ -1105,7 +1122,86 @@ def _solveset(f, symbol, domain, _check=False):
                 result = _result - singularities
 
     if _check:
-        if isinstance(result, ConditionSet):
+        if (isinstance(result,ConditionSet)) and \
+                    (not isinstance(result,list)):
+            x = Symbol('x',real=True)
+            f = f.subs({symbol: x})
+            if _is_lambert(f,x):
+
+                if result.has(cos,sin):
+                    if (result.has(exp)) or (f.has(sqrt(x))) or\
+                        f.has(Abs):
+                        return result
+                    elif not domain.is_subset(S.Integers) and domain.is_subset(S.Reals) and \
+                        ((result.has(cos) and not result.has(sin)) or \
+                            (result.has(sin) and not result.has(cos))) :
+                        f = solve(f, x)
+                        if f is not None and domain.is_subset(S.Reals):
+                            args = []
+                            j = -1
+                            for i in f:
+                                j += 1
+                                if not isinstance(i,int):
+                                    if not i.has(I) and i.is_real and domain.is_subset(S.Reals):
+                                        args.append(f[j])
+                                elif isinstance(i,int) and i.is_real and domain.is_subset(S.Reals):
+                                    args.append(f[j])
+                                if i.has(Dummy):
+                                    return EmptySet
+                            f = FiniteSet(*args)
+                            return f
+                        else:
+                            f = FiniteSet(*f)
+                            return f
+                elif domain.is_subset(S.Reals):
+                    if not any(i for i in f.atoms(Pow) if i.exp is S.Half) and f.count_ops() < 120:
+                        if not (f.has(exp) and f.has(Pow)):
+                            indls = _tsolve(f,x)
+                            if indls is None or indls == []:
+                                return result
+                            elif indls is not None:
+                                if len(indls) == 1:
+                                    if indls[0] == 0:
+                                        return result
+                                    if not indls[0].has(Symbol):
+                                        if not indls[0].is_real and domain.is_subset(S.Reals):
+                                            return S.EmptySet
```

Comment:
When your code ends being indented as much as this it's a clear sign that you need a better structure. Probably what is needed is to define some meaningful functions that can split this up.